src/audio/kpb.c

```diff
@@ -1404,48 +1404,97 @@ static void kpb_drain_samples(void *source, struct audio_stream *sink,
 	}
 }
 
+/**
+ * \brief Copy s16 data circular to linear
+ *
+ * \param[in] istream - input stream
+ * \param[in] dest - destination buffer
+ * \param[in] offset - start index as bytes to source buffer
+ * \param[in] samples - number of samples to copy
+ * \param[in] bytes_total - Requested copy size in bytes.
+ *
+ * \return none.
+ */
+static inline void buffer_s16(const struct audio_stream *istream, int16_t *dest, int offset,
+			      int bytes_total)
+{
+	int16_t *src = istream->r_ptr;
+	int16_t *dst = dest;
+	int bmax;
+	int b;
+	int bytes_copied = 0;
+
+	src = audio_stream_wrap(istream, src + (offset >> BYTES_TO_S16_SAMPLES));
+	while (bytes_copied < bytes_total) {
+		b = bytes_total - bytes_copied;
+		bmax = audio_stream_bytes_without_wrap(istream, src);
+		b = MIN(b, bmax);
+		memcpy(dst, src, b);
+		src = audio_stream_wrap(istream, src + (b >> BYTES_TO_S16_SAMPLES));
+		dst += b >> BYTES_TO_S16_SAMPLES;
+		bytes_copied += b;
+	}
```

Comment:
This looks like a very generic code optimization, we should make these common functions in stream.h as we will have other users..

Reply:
The number of bytes interface makes this and other optimized somewhat different than usual count of frames or samples parameter, and in this function the start offset. I can make a version to see how it looks like there.

Reply:
I've now moved the copy functions to audio_stream.h